### PR TITLE
Modified breeding to work per trait 

### PR DESCRIPTION
**Pull request by vedant-shroff**
Modified the genomes system to allow registering breeding algorithms per trait and a default breeding algorithm for the entity which will be used incase a breeding algorithm is not specified per trait. Also changed the traits to be stored in a linked hash map rather than a hash map so they are stored sequentially to maintain order. 
Added tests to check the working of Genomes while making modifications. 

**To-do list :**
- [x] Code Cleanup (major)
- [x] More tests to check breeding of multiple traits for the same entity

### CODE REVIEW

File: src/test/java/org/terasology/genome/GeneIndexGenomeMapTest.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+import org.terasology.genome.breed.BreedingAlgorithm;
+import org.terasology.genome.breed.MonoploidBreedingAlgorithm;
+import org.terasology.genome.breed.mutator.GeneMutator;
+import org.terasology.genome.breed.mutator.VocabularyGeneMutator;
+import org.terasology.genome.genomeMap.GeneIndexGenomeMap;
+import org.terasology.genome.genomeMap.SeedBasedGenomeMap;
+import org.terasology.registry.In;
+import org.terasology.world.WorldProvider;
+
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (e-aakash) [inline src/test/java/org/terasology/genome/GeneIndexGenomeMapTest.java]: \* imports should be avoided

File: src/main/java/org/terasology/genome/system/SimpleGenomeManager.java [hunk 2]
```diff
@@ -81,8 +98,27 @@ public boolean applyBreeding(EntityRef organism1, EntityRef organism2, EntityRef
...
+
+        while (geneIndex != genome1.genes.length()) {
+            geneIndices = genePropertyDefinitions.get(geneCounter).geneIndices;
+            breedingAlgorithm = genePropertyDefinitions.get(geneCounter++).breedingAlgorithm;
+            if (geneIndex + geneIndices.length < genome1.genes.length()) {
+                resultGenes += "" + breedingAlgorithm.produceCross(genome1.genes.substring(geneIndex,
+                        geneIndex + geneIndices.length), genome2.genes.substring(geneIndex,
+                        geneIndex + geneIndices.length));
+            } else {
+                System.out.println(resultGenes);
```
Comment (e-aakash) [inline src/main/java/org/terasology/genome/system/SimpleGenomeManager.java]: These (debug?) prints should be removed in final state of PR

File: src/test/java/org/terasology/genome/GeneIndexGenomeMapTest.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+                    }
+                });
+
+        GenomeDefinition genomeDefinition = new GenomeDefinition(breedingAlgorithm, genomeMap);
+        GeneIndexGenomeMap genomeMap1 = (GeneIndexGenomeMap) genomeDefinition.getGenomeMap();
+        assertEquals(genomeMap1.propertyDefinitionMap.toString(), genomeMap.propertyDefinitionMap.toString());
+        //System.out.println(genomeMap1.propertyDefinitionMap);
+        System.out.println(genomeMap1.getProperty("filling", "Aa", Integer.class));
+    }
+}
```
Comment (e-aakash) [inline src/test/java/org/terasology/genome/GeneIndexGenomeMapTest.java]: Newline at the end of the file is good practice

File: src/test/java/org/terasology/genome/GenomeNewTest.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+                        return (Character.isUpperCase(input.charAt(0)) ? 1 : 0);
+                    }
+                });
+
+        GenomeDefinition genomeDefinition = new GenomeDefinition(defaultBreedingAlgorithm, genomeMap);
+        GeneIndexGenomeMap genomeMap1 = (GeneIndexGenomeMap) genomeDefinition.getGenomeMap();
+        assertEquals(genomeMap1.propertyDefinitionMap.toString(), genomeMap.propertyDefinitionMap.toString());
+        assertEquals(genomeMap1.getProperty("height", "TT", Integer.class), 1);
+        assertEquals(genomeMap1.getProperty("height", "tt", Integer.class), 0);
+        System.out.println(genomeMap1.getProperty("height", "Tt", Integer.class));
```
Comment (e-aakash) [inline src/test/java/org/terasology/genome/GenomeNewTest.java]: Is there a specific reason for these outputs to console? If these are for confirming something, they can be converted to another assert statement?

File: src/test/java/org/terasology/genome/GenomeNewTest.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+                resultGenes += "" + breedingAlgorithm.produceCross(g1.substring(geneIndex,
+                        geneIndex + geneIndices.length), g2.substring(geneIndex,
+                        geneIndex + geneIndices.length));
+            } else {
+                System.out.println(resultGenes);
+                resultGenes += "" + breedingAlgorithm.produceCross(g1.substring(geneIndex), g2.substring(geneIndex));
+            }
+            geneIndex += geneIndices.length;
+        }
+        System.out.println(resultGenes);
```
Comment (e-aakash) [inline src/test/java/org/terasology/genome/GenomeNewTest.java]: This method is not testing anything as there are no assert statements.

**Review by e-aakash — CHANGES_REQUESTED** [review]
Added some review on the tests, and some style review.

File: src/test/java/org/terasology/genome/GenomeNewTest.java (hunk 4 above)
Comment (vedant-shroff) [reply src/test/java/org/terasology/genome/GenomeNewTest.java]: They were just used to check the values to confirm that the getProperty method was working correctly. Have replaced with assert statements

File: src/test/java/org/terasology/genome/GenomeNewTest.java (hunk 5 above)
Comment (vedant-shroff) [reply src/test/java/org/terasology/genome/GenomeNewTest.java]: The resultant genes are randomized according to the breeding algorithms and therefore the print statements were used to check if the result was correct or not. Just the working of this piece of code was enough for the test, but I have replaced it with an assert to check for an empty String at the end

**Review by e-aakash — APPROVED** [review]
Overall looks good. Tests working fine 👍

File: src/main/java/org/terasology/genome/genomeMap/SeedBasedGenomeMap.java [hunk 6]
```diff
@@ -35,38 +38,45 @@ public SeedBasedGenomeMap(int mapSeed) {
     /**
      * Add a new seed based (pseudo-random) property.
      *
-     * @param propertyName             The name of the property
-     * @param minGeneIndex             The minimum index of a which can affect the property
-     * @param maxGeneIndex             The maximum index of a gene that can affect the property
-     * @param codeLength               The maximum length of a gene sequence that can affect the property
-     * @param type                     The type of the property
+     * @param propertyName The name of the property
+     * @param minGeneIndex The minimum index of a which can affect the property
```
Comment (PS-Soundwave) [inline src/main/java/org/terasology/genome/genomeMap/SeedBasedGenomeMap.java]: Might as well fix this typo while we're here

File: src/main/java/org/terasology/genome/genomeMap/GeneIndexGenomeMap.java [hunk 7]
```diff
@@ -16,37 +16,54 @@
...
-import java.util.HashMap;
+import java.util.LinkedHashMap;
 import java.util.Map;
 
 /**
  * An implementation of GenomeMap.
  */
 public class GeneIndexGenomeMap implements GenomeMap {
-    private Map<String, GenePropertyDefinition> propertyDefinitionMap = new HashMap<>();
+    public Map<String, GenePropertyDefinition> propertyDefinitionMap = new LinkedHashMap<>();
```
Comment (skaldarnar) [inline src/main/java/org/terasology/genome/genomeMap/GeneIndexGenomeMap.java:30]: Why was this made public? This allows to modify or completely swap out the map, although we give specific acccessor methods below...

As this is public, it should also have Javadoc (which would be a good place to note down why can/should be done with it).

File: src/main/java/org/terasology/genome/genomeMap/GeneIndexGenomeMap.java [hunk 8]
```diff
@@ -16,37 +16,54 @@
...
+     * @param geneIndices The indices of the genes which the property depends
+     * @param type The type of the property
+     * @param breedingAlgorithm The breeding algorithm of the property
      * @param geneStringTransformation A function to transform the gene into a property value
-     * @param <T>                      The class used for the property value
+     * @param <T> The class used for the property value
      */
-    public <T> void addProperty(String propertyName, int[] geneIndices, Class<T> type, Function<String, T> geneStringTransformation) {
-        propertyDefinitionMap.put(propertyName, new GenePropertyDefinition<T>(geneIndices, type, geneStringTransformation));
+    public <T> void addProperty(String propertyName, int[] geneIndices, Class<T> type,
```
Comment (skaldarnar) [inline src/main/java/org/terasology/genome/genomeMap/GeneIndexGenomeMap.java:42]: Did you consider using `Name` instead of Strings to identify properties?

File: src/main/java/org/terasology/genome/genomeMap/GeneIndexGenomeMap.java [hunk 9]
```diff
@@ -16,37 +16,54 @@
...
+                geneStringTransformation));
     }
 
+    /**
+     * Get the breeding algorithm of the specified property.
+     *
+     * @param propertyName The name of the property to be added
+     */
+    public BreedingAlgorithm getPropertyBreedingAlgorithm(String propertyName) {
+        return propertyDefinitionMap.get(propertyName).breedingAlgorithm;
```
Comment (skaldarnar) [inline src/main/java/org/terasology/genome/genomeMap/GeneIndexGenomeMap.java:55]: This will easily throw a NPE if the property definitions does not contain the given property name. Consider changing the return type of this method to `Optional<BreedingAlgorithm>` and make the implementation null-safe, e.g., 

```java
return Optional.ofNullable(propertyDefinitionMap.get(propertyName))
            .map(GenePropertyDefinition::breedingAlgorithm);
```

File: src/test/java/org/terasology/genome/GeneIndexGenomeMapTest.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+/**
+ * Initial test setup based on working of Genomes before trait specific breeding algorithms were introduced
+ * this test is now obsolete/serves no purpose
+ */
+class GeneIndexGenomeMapTest {
+    @In
+    WorldProvider worldProvider;
+
+    @In
+    GenomeRegistry genomeRegistry;
```
Comment (skaldarnar) [inline src/test/java/org/terasology/genome/GeneIndexGenomeMapTest.java:45]: I'm surprised that this works (using `@In` annotations for tests o.O Did you actually run this, or are the fields used yet?

File: src/test/java/org/terasology/genome/GenomeNewTest.java [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+/**
+ * A test to check working of multiple traits with different breeding algorithms. Addition of multiple breeding
+ * algorithms was based on the structure of these tests.
+ */
+class GenomeNewTest {
+    @In
+    WorldProvider worldProvider;
+
+    @In
+    GenomeRegistry genomeRegistry;
```
Comment (skaldarnar) [inline src/test/java/org/terasology/genome/GenomeNewTest.java:56]: Same here, I don't think this works (not even when using the MTE). The tests probably pass nevertheless as these fields are never used. I'd be in favor of removing dead code (of preparing for some other changes, move it to another PR).

**Review by skaldarnar — COMMENTED** [review]
Only now found the time to have a quick look at this. Left a few comments and questions, mostly on code quality and API design. 
I'd appreciate if you'd consider these comments for upcoming PRs 😉

File: src/main/java/org/terasology/genome/genomeMap/GeneIndexGenomeMap.java (hunk 7 above)
Comment (vedant-shroff) [reply src/main/java/org/terasology/genome/genomeMap/GeneIndexGenomeMap.java:30]: It was made public so the Genome Manager can access the propertyDefinitionMap to get the appropriate traits in the order they are stored and access the details. It could be kept private and use a getter method to get the Map instead.

File: src/test/java/org/terasology/genome/GeneIndexGenomeMapTest.java (hunk 10 above)
Comment (vedant-shroff) [reply src/test/java/org/terasology/genome/GeneIndexGenomeMapTest.java:45]: I tried using it initially and it did not work. Took out the usage but forgot to remove the declarations, fixed :)

File: src/test/java/org/terasology/genome/GenomeNewTest.java (hunk 11 above)
Comment (vedant-shroff) [reply src/test/java/org/terasology/genome/GenomeNewTest.java:56]: same as above. will remove this

File: src/main/java/org/terasology/genome/genomeMap/GeneIndexGenomeMap.java (hunk 9 above)
Comment (vedant-shroff) [reply src/main/java/org/terasology/genome/genomeMap/GeneIndexGenomeMap.java:55]: This method is actually not used as it is sufficient to use the propertyDefinitionMap for the breeding algorithm of the traits, therefore I have removed it. But this is a fair point and I will keep it in mind for future work. For knowledge, is it preferred to use the Optional class for such a scenario or will a local variable with null check suffice?
